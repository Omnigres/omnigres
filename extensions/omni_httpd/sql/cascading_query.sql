CREATE TABLE routes (
   name text NOT NULL,
   query text NOT NULL,
   priority INT NOT NULL
);

INSERT INTO routes (name, query, priority) VALUES
  ('test', $$SELECT omni_httpd.http_response(body => 'test') FROM request WHERE request.path = '/test'$$, 1),
  ('ping', $$SELECT omni_httpd.http_response(body => 'pong') FROM request WHERE request.path = '/ping'$$, 1);

\pset format wrapped
\pset columns 80

SELECT omni_httpd.cascading_query(name, query ORDER BY priority DESC NULLS LAST) FROM routes GROUP BY priority ORDER BY priority DESC;

\pset format aligned

-- CTE handling

\pset format wrapped
\pset columns 80

SELECT omni_httpd.cascading_query(name, query ORDER by priority DESC NULLS LAST) FROM (
  VALUES
  ('test', $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'test') FROM request, Test WHERE request.path = '/test' and test.val = 1$$, 1),
  ('ping', $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'pong') FROM request, Test WHERE request.path = '/ping' and test.val = 1$$, 1))
  AS routes(name, query, priority) GROUP BY priority ORDER BY priority DESC;

 \pset format aligned