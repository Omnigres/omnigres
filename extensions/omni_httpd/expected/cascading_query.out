CREATE TABLE routes (
   name text NOT NULL,
   query text NOT NULL,
   priority INT NOT NULL
);
INSERT INTO routes (name, query, priority) VALUES
  ('test', $$SELECT omni_httpd.http_response(body => 'test') FROM request WHERE request.path = '/test'$$, 1),
  ('ping', $$SELECT omni_httpd.http_response(body => 'pong') FROM request WHERE request.path = '/ping'$$, 1);
\pset format wrapped
\pset columns 80
SELECT omni_httpd.cascading_query(name, query ORDER BY priority DESC NULLS LAST) FROM routes GROUP BY priority ORDER BY priority DESC;
                                cascading_query                                 
--------------------------------------------------------------------------------
 WITH test AS (SELECT omni_httpd.http_response(body := 'test') FROM request WHE.
.RE request.path = '/test'), ping AS (SELECT omni_httpd.http_response(body := '.
.pong') FROM request WHERE request.path = '/ping') SELECT * FROM test UNION ALL.
. SELECT * FROM ping WHERE NOT EXISTS (SELECT FROM test)
(1 row)

\pset format aligned
-- CTE handling
\pset format wrapped
\pset columns 80
SELECT omni_httpd.cascading_query(name, query ORDER by priority DESC NULLS LAST) FROM (
  VALUES
  ('test', $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'test') FROM request, Test WHERE request.path = '/test' and test.val = 1$$, 1),
  ('ping', $$WITH test AS (SELECT 1 AS val) SELECT omni_httpd.http_response(body => 'pong') FROM request, Test WHERE request.path = '/ping' and test.val = 1$$, 1))
  AS routes(name, query, priority) GROUP BY priority ORDER BY priority DESC;
                                cascading_query                                 
--------------------------------------------------------------------------------
 WITH __omni_httpd_test_test AS (SELECT 1 AS val), test AS (SELECT omni_httpd.h.
.ttp_response(body := 'test') FROM request, __omni_httpd_test_test test WHERE r.
.equest.path = '/test' AND test.val = 1), __omni_httpd_ping_test AS (SELECT 1 A.
.S val), ping AS (SELECT omni_httpd.http_response(body := 'pong') FROM request,.
. __omni_httpd_ping_test test WHERE request.path = '/ping' AND test.val = 1) SE.
.LECT * FROM test UNION ALL SELECT * FROM ping WHERE NOT EXISTS (SELECT FROM te.
.st)
(1 row)

 \pset format aligned
